var/spack/repos/builtin/packages/ipopt/package.py

```diff
@@ -109,14 +109,19 @@ def configure_args(self):
                     "--with-mumps-cflags=%s" % mumps_dir.include])
 
         if 'coinhsl' in spec:
+            hsl_ld_flags = '-ldl {0}'.format(spec['coinhsl'].libs.ld_flags)
```

Comment:
@ashermancinelli Ipopt's configure script adds `libdl` to its link lines when the lazy linear solver loader is used (see https://github.com/coin-or/Ipopt/blob/cfadeac29711a1c007752bbaa1ee33eea4f934f0/configure.ac#L469). I'm guessing that code uses `dlopen` and `dlsym` in the usual sort of way for shared library plugins. Adding `-ldl` should be moved outside this block because the lazy linear solver loader is enabled by default -- this Spack package has no variant to disable it -- independent of whether Ipopt is built with HSL.

I don't know off-hand whether there's a good way of exposing the `libdl` dependency to the concretizer.

Reply:
I think this has to do with coinhsl - it seems that coinhsl's build system will link to libdl sometimes (can't figure out when/why this happens) which then causes ipopt's configure to fail to find coinhsl (undefined symbols from libdl). Where do you suggest I put `-ldl`?

Reply:
@ashermancinelli If you can post some build logs, that would be a big help in understanding why CoinHSL needs `-ldl`, and why that flag needs to be added here, inside this particular block. It would also help to know which versions of CoinHSL and Ipopt you're trying to install.

The best case to make for why the flag should be added here is that there's something about linking CoinHSL specifically to Ipopt that requires this flag, but it's not obvious to me why that would be the case because the [Ipopt source code](https://github.com/coin-or/Ipopt/blob/303929d1d4103554d91ae6bbf299fd3f4a6a237a/src/Common/IpLibraryLoader.cpp#L62) uses it to load multiple solvers, not just CoinHSL.

From looking at the CoinHSL source code, I would guess that the best place to put this flag would be in the `libs` property of the CoinHSL package. Adding `-ldl` here would help all users of CoinHSL regardless of whether they're using it with Ipopt. I hesitate to post much source code here due to licensing issues, but in the CoinHSL 2014.01.10 source code, `loadmetis.c` requires `dlopen` on non-Windows operating systems. The version of CoinHSL you are using may have similar code in it.

The last possibility is that Ipopt requires this flag whether or not it's using CoinHSL. In recent versions, it does need this flag, and that flag is added whenever `--enable-linear-solver-loader` is passed to its `configure` script. If you're using a recent enough version of Ipopt (I think 3.12.13 or later), this flag would be on by default, so `-ldl` is already being added to the link line anyway. If for some reason it weren't, you'd add it to `LDFLAGS` inside the `ipopt` package. I originally thought this option was the best way forward, but based on the evidence, I no longer believe this option to be the best choice.